source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,348 @@
+
+  if (read == end || *read != '/') {
+    // Reject empty or relative paths
+    return {PathNormalizationResult::Action::Reject, UhvResponseCodeDetail::get().InvalidUrl};
+  }
+
+  ++read;
+  ++write;
+
+  // Path normalization is based on RFC 3986:
+  // https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
+  //
+  // SPELLCHECKER(off)
+  // path          = path-abempty    ; begins with "/" or is empty
+  //               / path-absolute   ; begins with "/" but not "//"
+  //               / path-noscheme   ; begins with a non-colon segment
+  //               / path-rootless   ; begins with a segment
+  //               / path-empty      ; zero characters
+  //
+  // path-abempty  = *( "/" segment )
+  // path-absolute = "/" [ segment-nz *( "/" segment ) ]
+  // path-noscheme = segment-nz-nc *( "/" segment )
+  // path-rootless = segment-nz *( "/" segment )
+  // path-empty    = 0<pchar>
+  // segment       = *pchar
+  // segment-nz    = 1*pchar
+  // segment-nz-nc = 1*( unreserved / pct-encoded / sub-delims / "@" )
+  //               ; non-zero-length segment without any colon ":"
+  //
+  // pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+  // SPELLCHECKER(on)
+  while (read != end) {
+    char ch = *read;
+    char prev = *std::prev(write);
+
+    switch (ch) {
+    case '%': {
+      auto decode_result = normalizeAndDecodeOctet(read, end);
+      switch (decode_result.result()) {
+      case PercentDecodeResult::Invalid:
+        ABSL_FALLTHROUGH_INTENDED;
+      case PercentDecodeResult::Reject:
+        // Reject the request
+        return {PathNormalizationResult::Action::Reject, UhvResponseCodeDetail::get().InvalidUrl};
+
+      case PercentDecodeResult::Normalized:
+        // Valid encoding but outside the UNRESERVED character set. The encoding was normalized to
+        // UPPERCASE and the octet must not be decoded. Copy the normalized encoding.
+        *write++ = *read++;
+        *write++ = *read++;
+        *write++ = *read++;
+        break;
+
+      case PercentDecodeResult::DecodedRedirect:
+        // The encoding was properly decoded but, based on the config, the request should be
+        // redirected to the normalized path.
+        redirect = true;
+        ABSL_FALLTHROUGH_INTENDED;
+      case PercentDecodeResult::Decoded:
+        // The encoding was decoded. Store the decoded octet in the last character of the percent
+        // encoding (read[2]) so it will be processed in the next iteration. We can safely advance
+        // 2 positions since we know that the value was correctly decoded.
+        std::advance(read, 2);
+        *read = decode_result.octet();
+      }
+      break;
+    }
+
+    case '.': {
+      if (prev == '/') {
+        // attempt to read ahead 2 characters to see if we are in a "./" or "../" segment.
+        const auto dot_segment = path_view.substr(std::distance(begin, read), 3);
+        if (absl::StartsWith(dot_segment, "./") || dot_segment == ".") {
```

Comment:
The problem here is that it would miss sequences like `/..%2f` (when decoding of %2f is enabled) or /%2E/ (. is in the unreserved set, so %2E should be decoded).

My overall comment here, is that we should trade some efficiency for clarity and split this big loop that does everything into 3 smaller consecutive loops:

1. decode UNRESERVED and %2F if configured and normalize case for other %2F
2. merge adjacent slashes
3. collapse dot and dot segements.

In this way I think it is easier to reason about various edge cases.

Reply:
Normalization is now 3 passes: decode, merge slashes (if configured), and collapse dot segments. The big change here is that the final path is not validated, which is fine because that is performed further on in the UHV process.